esp-hal-common/src/rtc_cntl.rs

```diff
@@ -1,46 +1,597 @@
-use crate::pac::RTC_CNTL;
+use fugit::{MegahertzU32, MicrosDurationU64};
 
-pub struct RtcCntl {
-    rtc_cntl: RTC_CNTL,
+use embedded_hal::watchdog::{Watchdog, WatchdogDisable, WatchdogEnable};
+
+use crate::{clock::Clock, clock::XtalClock, pac::RTC_CNTL, pac::TIMG0};
+
+#[cfg(not(feature = "esp32"))]
+use crate::efuse::Efuse;
+
+use crate::rom::esp_rom_delay_us;
+
+#[cfg_attr(feature = "esp32", path = "rtc/esp32.rs")]
+#[cfg_attr(feature = "esp32s2", path = "rtc/esp32s2.rs")]
+#[cfg_attr(feature = "esp32s3", path = "rtc/esp32s3.rs")]
+#[cfg_attr(feature = "esp32c3", path = "rtc/esp32c3.rs")]
+mod rtc;
+
+#[allow(unused)]
+#[derive(Debug, Clone, Copy)]
+/// RTC SLOW_CLK frequency values
+pub(crate) enum RtcFastClock {
+    /// Main XTAL, divided by 4
+    RtcFastClockXtalD4 = 0,
+    /// Internal 8 MHz RC oscillator
+    RtcFastClock8m = 1,
 }
 
-impl RtcCntl {
+impl Clock for RtcFastClock {
+    fn frequency(&self) -> MegahertzU32 {
+        match self {
+            RtcFastClock::RtcFastClockXtalD4 => MegahertzU32::Hz(40_000_000 / 4),
+            RtcFastClock::RtcFastClock8m => MegahertzU32::Hz(8_500_000),
+        }
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug, Clone, Copy)]
+/// RTC SLOW_CLK frequency values
+pub(crate) enum RtcSlowClock {
+    /// Internal 150 KHz RC oscillator
+    RtcSlowClockRtc = 0,
+    /// External 32 KHz XTAL
+    RtcSlowClock32kXtal = 1,
+    /// Internal 8 MHz RC oscillator, divided by 256
+    RtcSlowClock8mD256 = 2,
+}
+
+impl Clock for RtcSlowClock {
+    fn frequency(&self) -> MegahertzU32 {
+        match self {
+            RtcSlowClock::RtcSlowClockRtc => MegahertzU32::Hz(1_500_000),
```

Comment:
```suggestion
            RtcSlowClock::RtcSlowClockRtc => MegahertzU32::Hz(150_000),
```
The frequency here should be 150 kHz. I'll push a fix for it.
Once I fixed this one, I noticed that the `rtc_watchdog` demo application failed and this was due to using `MegahertzU32` to hold sub-MHz frequencies, which resulted in the value being truncated.
Then I changed the `Clock` trait to use `HertzU32` instead, which impacted other drivers which also rely on clocks.